### DISCUSSION TITLE
Why is gVisor made as the default TUN mode?

### DISCUSSION THREAD

**Discussion by a2232189** [SagerNet/SagerNet#355] [Q&A]
1. What is the use of gVisor/System TUN in SagerNet? 
2. Android VPN service is kind of a TUN if I remember correctly. What's the difference between the role of Android VPN service and gVisor/System options in the settings?
3. May I know the pros and cons of gVisor compared with system tun mode? 

Thanks.

**Comment by a2232189** [comment]
[It was made because it was always said that Clash for Android was faster, and CFA uses a similar implementation (mostly derived from clash-premium's system stack tun).](https://github.com/SagerNet/SagerNet/discussions/327#discussioncomment-1819976)

Why was Clash faster than Sager? 

 Is this gVisor option an attempt to make Sager faster?

**Reply by klzgrad** [reply]
> Why was Clash faster than Sager?

There's no such claim.

> the pros and cons of gVisor compared with system tun mode?

I haven't seen test data proving either way. If the dev has any, it should be documented.

**Reply by perdakovich** [reply]
I'm interested in the pros and cons of gVisor compared with system tun mode too.

**Reply by klzgrad** [reply]
[The "system tun mode" is also known as double-NAT](https://github.com/MatsuriDayo/NekoBoxForAndroid/issues/122)

**Reply by perdakovich** [reply]
> @klzgrad 
> [The "system tun mode" is also known as double-NAT](https://github.com/MatsuriDayo/NekoBoxForAndroid/issues/122)

I had to use a translator, but that explained a lot, thanks. 
I guess it's the same on GNU/Linux systems - better to use the System tcp stack?

UPD: Also what is VpnService? Is it some android-only thing, and if so, then how it works on GNU/Linux PCs?

**Reply by klzgrad** [reply]
The entire detour is because the user generally has no root on Android but they need some kind of transparent proxying. On Linux you go for iptables.
